### PR TITLE
TaskHeader and friends

### PR DESCRIPTION
**Pull request by jiivan**
fixes: https://github.com/golemfactory/golem/issues/3557

### CODE REVIEW

**Comment by etam** [comment]
hint for reviewing: `git diff -M30% master..origin/task_header` shows a rename and changes between `golem_messages/datastructures.py` and `golem_messages/datastructures/__init__.py` instead of showing files deleted and added.

File: golem_messages/datastructures/__init__.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+        if callable(validator):
+            validator(value=value)  # noqa pylint:disable=not-callable
+
+        super().__setitem__(key, value)
+
+
+class StringEnum(str, enum.Enum):
+
+    # pylint: disable=no-self-argument
```
[the remark this answers is not in the document]
Comment (jiivan) [reply golem_messages/datastructures/__init__.py]: Outside of this PR scope

File: golem_messages/datastructures/__init__.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+class Container:
+    """Container implementation supporting serialization and deserialization"""
+    # Values in __slots__ are iterables of validators.
+    # Each validator is called with (field_name, value)
+    # See: deserialize_slot()
+    __slots__: typing.Dict[str, typing.callable] = {}
+    # List of required slot names
+    REQUIRED: typing.FozenSet[str] = frozenset()
+    # Default values for slots. See: load_slots
+    DEFAULTS: typing.Dict[str, typing.ANY] = {}
```
Comment (etam) [inline golem_messages/datastructures/__init__.py]: Gotcha! Values of `DEFAULTS` are not any type, but callables, that return the default value when called without arguments.

File: golem_messages/datastructures/__init__.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+    # pylint: enable=no-self-argument
+
+
+class Container:
+    """Container implementation supporting serialization and deserialization"""
+    # Values in __slots__ are iterables of validators.
+    # Each validator is called with (field_name, value)
+    # See: deserialize_slot()
+    __slots__: typing.Dict[str, typing.callable] = {}
```
Comment (etam) [inline golem_messages/datastructures/__init__.py]:
```suggestion
    __slots__: typing.Dict[str, typing.Callable[[str, typing.Any], None]] = {}
```

File: golem_messages/datastructures/tasks.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        'deadline': (
+            validators.validate_float,
+            functools.partial(
+                _fail_if,
+                check=lambda x: x > time.time(),
+                fail_msg="Deadline already passed",
+            ),
+        ),
+        'subtask_timeout': (
+            validators.validate_float,
```
Comment (etam) [inline golem_messages/datastructures/tasks.py]: I think that timeouts should be ints. See my comments at https://github.com/golemfactory/golem/pull/3655

**Pull request merged**